### PR TITLE
Initial skeleton for C++ example/exercise code

### PR DESCRIPTION
**Pull request by alexdewar**
The code I've used for the first example/exercise is the one used by @lkr-ind for Chapter 2 of the course. I had to reorganise things a bit to make it fit into a CMake project, but the code itself should be the same.

I've also added a top-level CMakeLists.txt file to the project. One reason is that it's useful for CI purposes, but a top-level CMakeLists.txt file is also needed for IDEs like Visual Studio Code to correctly load the project for things like intellisense etc. Rather pleasingly, if you do this then the test suite is also nicely integrated with the IDE:

image

The common CMake code currently lives in its own file under the `cmake/` folder, but if students want to make their own project they would have to copy and paste its contents into their own CMakeLists.txt

I've also enabled CI for the repo so that we can check that all the example code compiles on Windows, Linux and macOS and that the files are properly formatted etc. Currently I'm *not* running the tests, as I wasn't sure whether we'd want some failing tests in the repository as an exercise. In that case we could mark the tests as expected to fail (assuming you can do this with GoogleTest), but that might confuse students.

I've stuck some text in a readme, but some/all of it could also just go directly into the course material, depending on how we want to organise things.

Close https://github.com/ImperialCollegeLondon/unit_testing_Cpp/issues/7

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+      - name: Set up MSVC (Windows)
+        if: matrix.os == 'windows-latest'
+        uses: ilammy/msvc-dev-cmd@v1
+        with:
+          arch: x64
+
+      - name: Install dependencies (macOS)
+        if: matrix.os == 'macos-latest'
+        run: brew install cmake ninja
```
Comment (dalonsoa) [inline .github/workflows/ci.yml:48]: I think `cmake` is included by default in macos runners: https://github.com/actions/runner-images/blob/main/images/macos/macos-13-Readme.md

Comment (dalonsoa) [reply .github/workflows/ci.yml:48]: What's `Ninja`, by the way? Would the potential target audience of the course know about it?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+Then you can build using CMake in the usual way:
+
+```sh
+cmake -B build
+cmake --build build/
+```
+
+## Running tests from the command line
+
+Once you have built your project, you can then run the tests by using [CTest], which is
```
Comment (dalonsoa) [inline README.md:46]: It is unclear to me what's the difference between `gtest` and `CTest`. Given that we talk about `gtest` in the course itself, it might confusing for students to come here and find out they need to use `CTest` instead unless some explanation is given.

**Review by dalonsoa — APPROVED** [review]
I've made a couple of comments but, except for some clarification in the README, it looks good!

**Comment by dalonsoa** [comment]
Close https://github.com/ImperialCollegeLondon/unit_testing_Cpp/issues/7

File: .github/workflows/ci.yml (hunk 1 above)
Comment (alexdewar) [reply .github/workflows/ci.yml]: I deliberately omitted mention of it from the README, because I don't want students to try to use it. They should be able to build things just by following the instructions.

It's basically a cross-platform version of `make` that's fast and has a less verbose output. It's useful for the CI system because it runs in parallel by default and also works with Windows. In the past I've found that CMake generates Visual Studio project files on Windows by default, but maybe that's no longer true (or maybe it would work from the command line anyway).

https://ninja-build.org/

File: README.md (hunk 2 above)
Comment (alexdewar) [reply README.md:46]: I probably have confused matters a bit here. I'm still using GoogleTest (`gtest`) in this repo though!

CTest is a part of CMake for integrating unit tests, so you don't have to execute the test program directly and your IDE can easily find it etc. I think it might also be a test framework in its own right, but that's not how we're using it here.

**Comment by jamesturner246** [comment]
Is it possible to have two sets of testsuites with cmake? A hidden one for us, which checks all the code compiles, and the tests pass/xfail, and one for exercises which includes purposely failing tests, etcetera?

File: .github/workflows/pre-commit_autoupdate.yml
Comment (jamesturner246) [inline .github/workflows/pre-commit_autoupdate.yml:1]: That's cool. Did you write these workflows yourself, or get these from the workflow 'marketplace' thing?

File: README.md (hunk 2 above)
Comment (jamesturner246) [reply README.md:52]: Maybe mentioning you can call the compiled google tests directly, without using (or if you don't have access to) CTest.

File: .github/workflows/pre-commit_autoupdate.yml
Comment (alexdewar) [reply .github/workflows/pre-commit_autoupdate.yml:1]: I mostly copy-pasted from the one in Health-GPS but tweaked it a little.

File: README.md (hunk 2 above)
Comment (alexdewar) [reply README.md:52]: @jamesturner246 Good point. In a way, using CTest rather than invoking the test binary directly means you end up glossing over that detail, which is probably an important one. I was thinking that maybe we could talk students through this in the course though, rather than in the instructions.

E.g., something like:

> After you compile your program, you'll notice that there's another binary called `Chapter2_test` in folder [whatever]. If you run this, you can see whether your tests are passing or failing.
>
> As we are using CMake though, you can also just run `ctest`, which will run your test binary (or binaries) for the project -- whatever they are called -- and display a summary of the results. This is also the way that many IDEs will run your tests, if they have this feature.

What does everyone think?

File: src/Chapter2/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,13 @@
+cmake_minimum_required(VERSION 3.11)
+project(Chapter2)
+include(../../cmake/UnitTestingCourseCommon.cmake)
+
+add_executable(${PROJECT_NAME}_program program.cpp fibonacci.cpp)
+add_executable(${PROJECT_NAME}_test test.cpp fibonacci.cpp)
+target_link_libraries(
+    ${PROJECT_NAME}_test
+    GTest::gtest_main
```
Comment (alexdewar) [inline src/Chapter2/CMakeLists.txt:9]: The course doesn't use `gtest_main`, so I'll update this.

Comment (lkr-ind) [reply src/Chapter2/CMakeLists.txt:9]: I have just started to see your PR. The course uses `gtest_main` in the next chapters. Please take a look at my new PR (https://github.com/ImperialCollegeLondon/unit_testing_Cpp/pull/14) for test fixtures. We might need this library.

Comment (alexdewar) [reply src/Chapter2/CMakeLists.txt:9]: Ah cool, ok. Well I guess I'll just update it so that it's in sync with the code you present in the course, when your PRs are merged.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+      - name: Set up MSVC (Windows)
+        if: matrix.os == 'windows-latest'
+        uses: ilammy/msvc-dev-cmd@v1
+        with:
+          arch: x64
+
+      - name: Install dependencies (macOS)
+        if: matrix.os == 'macos-latest'
+        run: brew install ninja
```
Comment (lkr-ind) [inline .github/workflows/ci.yml:48]: My question might be wrong. Please correct me if it is the case.

I guess, all 3 actions viz: Install dependencies (Ubuntu),  Set up MSVC (Windows) and Install dependencies (macOS) will be checked for their if condition and only one will execute depending on the system. 

Is it possible, to rearrange it slightly so that name could be like `Install dependencies` and inside that we checl all 3 conditions. 

In short, I am asking, can we put all conditions in one subheading (-name tag)?

**Review by lkr-ind — APPROVED** [review]
Thanks @alexdewar  for your work. This not only looks good but will add positive value to the course, its organization and will help the students a lot.

If you have some time next week, it will be good if you can integrate other code exercises as well. I am happy to talk to you for the same.

File: README.md (hunk 2 above)
Comment (jamesturner246) [reply README.md:52]: Fair enough, that also works

File: .github/workflows/ci.yml (hunk 4 above)
Comment (alexdewar) [reply .github/workflows/ci.yml:48]: It's a good question. I don't know, honestly. The syntax for these files is a little ugly, but I agree it would be nicer. I'll take a look on Monday.

**Comment by alexdewar** [comment]
Thanks for feedback @lkr-ind. I've opened a new issue to add the rest of the code: #2

File: .github/workflows/ci.yml (hunk 4 above)
Comment (dalonsoa) [reply .github/workflows/ci.yml:48]: I would disagree here. I think that having an step per platform with a clear, simple condition of when to run that step is a cleaner approach than trying to put everything in a single step - even if that were possible - just to have a single “Install dependencies” section.

Comment (alexdewar) [reply .github/workflows/ci.yml:48]: Ok, maybe I'll just leave as is then.
